runtime/src/main/cpp/Atomic.cpp

```diff
@@ -62,9 +82,45 @@ void Kotlin_AtomicReference_checkIfFrozen(KRef value) {
     }
 }
 
-KRef Kotlin_AtomicReference_compareAndSwap(KRef thiz, KRef expectedValue, KRef newValue) {
+// TODO: maybe those two belongs to Memory.cpp, as somewhat depends on ARC/GC semantics.
+OBJ_GETTER(Kotlin_AtomicReference_compareAndSwap, KRef thiz, KRef expectedValue, KRef newValue) {
     Kotlin_AtomicReference_checkIfFrozen(newValue);
-    return compareAndSwapImpl(thiz, expectedValue, newValue);
+    if (expectedValue == newValue) {
+        RETURN_OBJ(expectedValue);
+    }
+    // See Kotlin_AtomicReference_get() for explanations, why locking is needed.
+    lock(thiz);
+    // As we lock [thiz], atomicity here is not really important.
+    KRef old = compareAndSwapImpl(thiz, expectedValue, newValue);
+    unlock(thiz);
+    if (old == expectedValue) {
+        // CAS to the new value was successful, we transfer ownership of [expectedValue] to the caller.
+        // No need to update the reference counter, as before @AtomicReference was holding the reference,
+        // and now result holds it. Also note, that object referred by [expectedValue] cannot be freed,
+        // as it is guaranteed to be held by the caller (and by [thiz]).
+        *OBJ_RESULT = old;
+        return old;
+    } else {
+        // On this path we just create an additional reference to the [expectedValue], held by caller anyway.
```

Comment:
Data race can also occur here, if other thread performs CAS between `unlock` and this line.
And the comment seems to be obsolete.